README
XBill 2.1

Welcome to xbill...

Ever get the feeling that nothing is going right?  You're a sysadmin,
and someone's trying to destroy your computers.  The little people
running around the screen are trying to infect your computers with
Wingdows [TM], a virus cleverly  designed to resemble a popular
operating system.  Your objective is to click the mouse on them, ending
the potential threat.  If one of the people reaches a computer, it will
attempt to replace your operating system with the virus it carries. It
will then attempt to run off the screen with your vital software.  The
game ends when only 1 (or 0) of your computers are being productive.
Additionally, some computers are connected with network cables.  When
one computer on a network becomes infected, a spark will be sent down
the cable, and will infect the computer on the other end when it reaches
there.

Clicking the button on one of the little people will cause it to cry out
in pain and melt (id software eat your heart out!), dropping the stolen
os if it is carrying one.  If a computer is running Wingdows or is
temporarily off, the os can be dragged back to the computer (or
another computer compatible with that os).  To extinguish a spark drag
the bucket of water from the upper left corner onto it.

	The status bar at the bottom tells the following:
		Number of Bills on/off the screen
		Number of Computers running their OS/off/Wingdows
		Level
		Score

Options:
	-l <n>	Start at level n.

	--gui <gui>
		Use a specific front end.  The possible values are
		gtk, motif, and athena.  Note that some of these
		may not be compiled into the binary.

	-size <size>
		Play on a field of size x size, instead of the normal
		400x400.  <size> must be larger than 400.

	-v	Print version number and exit.

	-h	Print help and exit.

	When using the GTK gui, all standard GTK options are supported.
	When using the Athena or Motif GUI, all standard X Intrinsics
	options are supported.


Authors:
	Main Programmer:
		Brian Wellington <bwelling@xbill.org>
	Programming & graphics (v2.0 and earlier):
		Matias Duarte <matias@hyperimage.com>


ChangeLog
XBill history:

0.01 (7/21/94):  
	- Bill moves, and he's in color too.  You can even click on
		him and he excaims "OUCH!" or "SPLAT!"
	- window resizing is very wrong
	
0.02 (7/22/94):	
	- program no longer guaranteed to run predictably for 
		longer than nine years (sorry)
	- ridiculously improved code style (no really a lot!)
	- added linked list unit and support for multiple monsters
	- added support for computers
	- bill says "POW!!!!!!" on button 2
	- improved Makefile
	- remembered to include app-defaults file in tar

0.03 (7/24/94):
	- forgot to initialize a variable.  either the stupid aix compiler 
		initialized it anyway or we were really lucky testing it.
	- removed annoying flicker
	- has multiple computers on the screen with random OSes
	- has multiple bills moving on the screen in random directions
	- used sxpm to create smaller .xpm files, minor savings in memoy
		and executable size, easier to edit pixmaps
	- left button (button 1) removes bills from screen
	- right button (button 3) now resets computers to wingdows.

0.04 (7/25/94):
	- got masking to work (well, maybe not...)
	- bill now carries the wingdows box around
	- event handling/speed improved

0.1 (7/26/94):
	- masking really works now
	- fixed computer pixmap so it looks nice(r)
	- randomized computer layout now leaves room for bill to stand on right
	- support for bills walking in, standing at computers and walking out
	- bills walk towards a specific computer to plunder (primitive ai)
	- bill turns off computer, 'steals' native os and installs wingdows
	- count of bills remaining is kept: decreased when bill enters screen
					  - increased if bill leaves
	- bill is killable in all states
	- computer reset to native os on right button
	- cleaned up code and internal structures
	- event handling/speed really improved (optimized for the new code)
	- server lag problems identified
	- linked list unit fixed so that function to move between lists works
	- window resizing is still very wrong
	
0.2 (8/14/94)
	- programmers woke up after being asleep since 7/26.	
	- code structures were fixed very much (again)
	- bill faces in the direction he's walking
	- bill explodes in a gory shower of blood when clicked on (yeah!)
	- bill drops os when killed
	- the os stays on the ground when bill is killed if it had been stolen
	- the os bounces up and down while bill carries it
	- changed default cursor (matias will have to draw a better one later)
	- cursor changes during drag and drop operation
	- drag and drop works!  there's a little hack involved, but not 
		too bad...
	- bill now smart enough to notice that a computer is already
		corrupted and find a new one
	- game ends if all computers are lost

0.21 (8/15/94)
	- attempt to port code to athena widgets.  it compiles and
		runs w/o motif now (i hate C macros) a bit
	- improved the makefile.  an imakefile one day maybe...

0.22 (8/16/94)
	- when bill leaves the field, he drops the os he's carrying
	- status bar widget now does something

0.3 (8/23/94)
	- improved responsiveness a lot :)
	- menubar does something (new game, pause and quit buttons) if 
		motif is used.
	- levels now exist (barely)
	- yet more code cleanup

0.31 (8/24/94)
	- a few things depend on the level (freq. and number of appearing
		bills, number of computers, step size)
	- bug where a computer is ignored if off is removed
	- menu code written.
	- athena code started.  this works more or less, but not too well.
		i'd recommend using motif and/or a statically linked binary.

0.32 (9/18/94)
	- about box works.  help almost works also.
	- a bunch of bugs fixed
	- have an Imakefile now (thanks to Greg Seidman - anthro@cs.umd.edu)

0.9 (11/28/94 - a bit busier during school)
	- warp to arbitrary level works
	- game pauses when cursor leaves window
	- cursor warps to near upper left corner of opened dialog boxes
	- redefined both reset and drag to right button (button 3),
		added -1 option for people with a one button mouse.
		-1 may slow the game down a bit. 
	- added -l, -L to specify starting level
	- added -v for version info
	- improved pictures a lot
	- added confirmations for actions
	- game now pauses well
	- fixed level difficulty stuff
	- bill walks at two speeds
	- minimizes to an icon (a cool one too...)
	- scoring improved
	- window cannot be resized now, no matter how hard you try :)
	- much cooler animations for bill replacing os
	- added high score list
	- fixed lots of silly bugs
	- made program not die from colormap error (i hope)

1.0 (4/5/95-4/9/95)
	- remembered that the game's not done yet :)
	- replaced usleep() calls with XtAppAddTimeout().  should make
		the game more responsive
	- changed all popup windows to modal dialogs.  cleans up code
		a lot
	- removed Athena support, since it would take too long to finish.
	- made a couple pixmaps a bit smaller
	- changed colors to gray.  why not?
	- finally got everything to work with full colormaps
	- got rid of app-defaults file.  Not really used anyway
	- Only uses one button now (any button).  Easier to play.
	- Some Imakefile fixes

1.1 (4/16/95)
	- modularized widget code, and added Athena widget support (rewritten
		from scratch).  The game looks better with Motif, but that's
		not too important.
	- made pause button only active during game play
	- warping to a level starts a game if one isn't being played
	- fixed a potentially nasty bug that could cause bills to get lost
	- works in TrueColor now (I think - it works in 16 bit mode at least)

1.1a (9/22/95)
	- fixed bug where getopt() returned a char instead of an int.
	- fixed bug where hitting Enter when entering a high score
		corrupted the high score file.
	- don't fclose() nonexistant files any more.
	- added BSD as another operating system (patches by Joerg Wunsch,
		joerg_wunsch@uriah.heep.sax.de)
	- it was really easy to play forever by guarding one computer.  Now
		you lose when there's only one computer left running its
		normal OS.

OS.c
#include "types.h"
#include "util.h"

#include "OS.h"
#include "UI.h"

#define MIN_PC 6		/* OS >= MIN_PC means the OS is a PC OS */

static const char *osname[] = {"wingdows", "apple", "next", "sgi", "sun",
			       "palm", "os2", "bsd", "linux", "redhat", "hurd"};
#define NUM_OS (sizeof(osname) / sizeof(osname[0]))

static Picture *os[NUM_OS];		/* array of OS pictures*/
static MCursor *cursor[NUM_OS];		/* array of OS cursors (drag/drop) */


void
OS_load_pix() {
	unsigned int i;
	for (i = 0; i < NUM_OS; i++) {
		UI_load_picture(osname[i], 1, &os[i]);
		if (i != 0)
			UI_load_cursor(osname[i], CURSOR_OWN_MASK, &cursor[i]);
	}
}

void
OS_draw(int index, int x, int y) {
	UI_draw(os[index], x, y);
}

int
OS_width() {
	return UI_picture_width(os[0]);
}

int
OS_height() {
	return UI_picture_height(os[0]);
}

void
OS_set_cursor(int index) {
	UI_set_cursor(cursor[index]);
}

int
OS_randpc() {
	return (RAND(MIN_PC, NUM_OS - 1));
}

int
OS_ispc(int index) {
	return (index >= MIN_PC);
}


strings.h
#ifndef STRINGS_H
#define STRINGS_H

/*
 * Dialog boxes
 */

#define newgame_dialog_str "New Game?"
#define pause_dialog_str "Game paused.  Press Continue to continue."
#define warp_dialog_str "Warp to level?"
#define quit_dialog_str "Quit Game?"

#define story_dialog_str \
"The Story:\n\
\n\
Yet again, the fate of the world rests\n\
in your hands!  An evil computer hacker,\n\
known only by his handle 'Bill', has\n\
created the ultimate computer virus.  A\n\
virus so powerful that it has the power\n\
to transmute an ordinary computer into\n\
a toaster oven.  (oooh!) 'Bill' has\n\
cloned himself into a billion-jillion\n\
micro-Bills.  Their sole purpose is to\n\
deliver the nefarious virus, which has\n\
been cleverly diguised as a popular\n\
operating system.\n\
\n\
As System Administrator/Exterminator,\n\
your job is to keep Bill from succeeding\n\
at his task."

#define rules_dialog_str \
"The Rules:\n\
\n\
xBill has been painstakingly designed and\n\
researched in order to make it as easy to use\n\
for the whole family as it is for little Sally.\n\
Years - nay - days of beta testing and \n\
consulting with the cheapest of human interface\n\
designers have resulted in a game that is easy\n\
to use, yet nothing at all like a Macintosh.\n\
\n\
I.   Whack the Bills (click)\n\
II.  Restart the computer (click)\n\
III. Pick up stolen OSes & return(drag)\n\
     them to their respective computers\n\
IV.  Drag the bucket to extinguish sparks\n\
V.   Scoring is based on total uptime,\n\
     with bonuses for killing Bills.\n\
\n\
As for the rest, you can probably figure\n\
it out.  We did, so it can't be too hard."

#define endgame_dialog_str \
"Module xBill has caused a segmentation fault\n\
at memory address 097E:F1A0.  Core dumped.\n\
\n\
We apologize for the inconvenience."

#define entername_dialog_str "You earned a high score.\nEnter your name:"


/*
 * Menus
 */
#define newgame_menu_str "New Game"
#define pause_menu_str "Pause Game"
#define warp_menu_str "Warp to level..."
#define highscore_menu_str "View High Scores"
#define quit_menu_str "Quit Game"

#define story_menu_str "Story of xBill"
#define rules_menu_str "Rules"
#define about_menu_str "About xBill"

#define score_menu_str "Score"
#define endgame_menu_str "End Game"
#define entername_menu_str "Enter Name"

#endif


Computer.c
#include <stdlib.h>

#include "types.h"
#include "util.h"

#include "Bill.h"
#include "Computer.h"
#include "Horde.h"
#include "Network.h"
#include "Game.h"
#include "OS.h"
#include "UI.h"

#define OS_OFFSET 4			/* offset of screen from 0,0 */
#define BORDER(size) (size / 10)	/* at least this far from a side */

#define MIN_PC 6		/* type >= MIN_PC means the computer is a PC */

static const char *cpuname[] = {"toaster", "maccpu", "nextcpu", "sgicpu",
				"suncpu", "palmcpu", "os2cpu", "bsdcpu"};

#define NUM_SYS (sizeof(cpuname) / sizeof(cpuname[0]))

static Picture *pictures[NUM_SYS];		/* array of cpu pictures */
static int width, height;


static int
determineOS(Computer *computer) {
	if (computer->type < MIN_PC)
		return computer->type;
	else
		return OS_randpc();
}

int
Computer_setup(Computer *computer, int index) {
	int j, counter = 0, flag;
	int x, y;
	int screensize = Game_screensize();
	int border = BORDER(screensize);
	do {
		if (++counter > 4000)
			return 0;
		x = RAND(border, screensize - border - width);
		y = RAND(border, screensize - border - height);
		flag = 1;
		/* check for conflicting computer placement */
		for (j = 0; j < index && flag; j++) {
			Computer *c = Network_get_computer(j);
			int twidth = width - BILL_OFFSET_X + Bill_width();
			if (UI_intersect(x, y, twidth, height,
					 c->x, c->y, twidth, height))
				flag = 0;
		}
	} while (!flag);
	computer->x = x;
	computer->y = y;
	computer->type = RAND(1, NUM_SYS - 1);
	computer->os = determineOS(computer);
	computer->busy = 0;
	computer->stray = NULL;
	return 1;
}

int
Computer_on(Computer *computer, int locx, int locy) {
	return (abs(locx - computer->x) < width &&
		abs(locy - computer->y) < height);
}

int
Computer_compatible(Computer *computer, int system) {
	return (computer->type == system ||
		(computer->type >= MIN_PC && OS_ispc(system)));
}

void
Computer_draw(Computer *computer) {
	UI_draw(pictures[computer->type], computer->x, computer->y);
	if (computer->os != OS_OFF)
		OS_draw(computer->os,
			computer->x + OS_OFFSET, computer->y + OS_OFFSET);
}

void
Computer_load_pix() {
	unsigned int i;
	for (i = 0; i < NUM_SYS; i++)
		UI_load_picture(cpuname[i], 1, &pictures[i]);
	width = UI_picture_width(pictures[0]);
	height = UI_picture_height(pictures[0]);
}

int
Computer_width() {
	return width;
}

int
Computer_height() {
	return height;
}


Game.c
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <time.h>

#include "types.h"
#include "util.h"

#include "Bill.h"
#include "Bucket.h"
#include "Computer.h"
#include "Cable.h"
#include "Game.h"
#include "Horde.h"
#include "Network.h"
#include "OS.h"
#include "Scorelist.h"
#include "Spark.h"
#include "UI.h"

#define SCREENSIZE 400

/* Game states */
#define STATE_PLAYING 1
#define STATE_BETWEEN 2
#define STATE_END 3
#define STATE_WAITING 4

/* Score related constants */
#define SCORE_ENDLEVEL -1
#define SCORE_BILLPOINTS 5

static unsigned int state;
static int efficiency;
static int score, level, iteration;
static Picture *logo, *icon, *about;
static MCursor *defaultcursor, *downcursor;
static Bill *grabbed;
static const char *gui;
static int screensize = SCREENSIZE;

static void
setup_level(int newlevel) {
	level = newlevel;
	Horde_setup();
	grabbed = NULL;
	UI_set_cursor(defaultcursor);
	Network_setup();
	iteration = 0;
	efficiency = 0;
}

void
Game_start(int newlevel) {
	state = STATE_PLAYING;
	score = 0;
	UI_restart_timer();
	UI_set_pausebutton(1);
	setup_level(newlevel);
}

void
Game_quit() {
	Scorelist_write();
	exit(0);
}

static void
update_info(void) {
	char str[80];
	int on_screen = Horde_get_counter(HORDE_COUNTER_ON);
	int off_screen = Horde_get_counter(HORDE_COUNTER_OFF);
	int base = Network_get_counter(NETWORK_COUNTER_BASE);
	int off = Network_get_counter(NETWORK_COUNTER_OFF);
	int win = Network_get_counter(NETWORK_COUNTER_WIN);
	int units = Network_num_computers();
	sprintf(str, "Bill:%d/%d  System:%d/%d/%d  Level:%d  Score:%d",
		on_screen, off_screen, base, off, win, level, score);
	UI_draw_str(str, 5, screensize - 5);
	efficiency += ((100 * base - 10 * win) / units);
}

void
Game_warp_to_level(int lev) {
	if (state == STATE_PLAYING) {
		if (lev <= level)
			return;
		setup_level(lev);
	}
	else {
		if (lev <= 0)
			return;
		Game_start(lev);
	}
}

void
Game_add_high_score(const char *str) {
	Scorelist_recalc(str, level, score);
}

void
Game_button_press(int x, int y) {
	int counter;

	if (state != STATE_PLAYING)
		return;
	UI_set_cursor(downcursor);

	if (Bucket_clicked(x, y)) {
		Bucket_grab(x, y);
		return;
	}

	grabbed = Horde_clicked_stray(x, y);
	if (grabbed != NULL) {
		OS_set_cursor(grabbed->cargo);
		return;
	}

	counter = Horde_process_click(x, y);
	score += (counter * counter * SCORE_BILLPOINTS);
}

void
Game_button_release(int x, int y) {
	int i;
	UI_set_cursor(defaultcursor);

	if (state != STATE_PLAYING)
		return;

	if (grabbed == NULL) {
		Bucket_release(x, y);
		return;
	}

	for (i = 0; i < Network_num_computers(); i++) {
		Computer *computer = Network_get_computer(i);

		if (Computer_on(computer, x, y) &&
		    Computer_compatible(computer, grabbed->cargo) &&
		    (computer->os == OS_WINGDOWS || computer->os == OS_OFF)) {
			int counter;

			Network_inc_counter(NETWORK_COUNTER_BASE, 1);
			if (computer->os == OS_WINGDOWS)
				counter = NETWORK_COUNTER_WIN;
			else
				counter = NETWORK_COUNTER_OFF;
			Network_inc_counter(counter, -1);
			computer->os = grabbed->cargo;
			Horde_remove_bill(grabbed);
			grabbed = NULL;
			return;
		}
	}
	Horde_add_bill(grabbed);
	grabbed = NULL;
}

static void
draw_logo(void) {
	UI_clear();
	UI_draw(logo,
		(screensize - UI_picture_width(logo)) / 2,
		(screensize - UI_picture_height(logo)) / 2);
}

void
Game_update() {
	char str[40];

	switch (state) {
	case STATE_PLAYING:
		UI_clear(); 
		Bucket_draw();
		Network_update();
		Network_draw();
		Horde_update(iteration);
		Horde_draw();
		update_info();
		if (Horde_get_counter(HORDE_COUNTER_ON) +
		    Horde_get_counter(HORDE_COUNTER_OFF) == 0) {
			score += (level * efficiency / iteration);
			state = STATE_BETWEEN;
		}
		if ((Network_get_counter(NETWORK_COUNTER_BASE) +
		     Network_get_counter(NETWORK_COUNTER_OFF)) <= 1)
			state = STATE_END;
		break;
	case STATE_END:
		UI_set_cursor(defaultcursor);
		UI_clear();
		Network_toasters();
		Network_draw();
		UI_refresh();
		UI_popup_dialog(DIALOG_ENDGAME);
		if (Scorelist_ishighscore(score)) {
			UI_popup_dialog(DIALOG_ENTERNAME);
			Scorelist_update();
		}
		UI_popup_dialog(DIALOG_HIGHSCORE);
		draw_logo();
		UI_kill_timer();
		UI_set_pausebutton(0);
		state = STATE_WAITING;
		break;
	case STATE_BETWEEN:
		UI_set_cursor(defaultcursor);
		sprintf(str, "After Level %d:\nScore: %d", level, score);
		UI_update_dialog(DIALOG_SCORE, str);
		UI_popup_dialog(DIALOG_SCORE);
		state = STATE_PLAYING;
		setup_level(++level);
		break;
	}
	UI_refresh();
	iteration++;
}

int
Game_score() {
	return score;
}

int
Game_level() {
	return level;
}

int
Game_screensize() {
	return screensize;
}

double
Game_scale(int dimensions) {
	double scale = (double)screensize / SCREENSIZE;
	double d = 1;
	for ( ; dimensions > 0; dimensions--)
		d *= scale;
	return (d);
}

/*
 * Note - don't use getopt, since it might reorder the args or do something
 * that the UI-specific argument parser doesn't like.
 */
static void
parse_args(int argc, char **argv) {
	char *s;
	char *endp;
	int i;

	for (i = 1; i < argc; i++) {
		if (strncasecmp(argv[i], "-l", 2) == 0) {
			if (strlen(argv[i]) == 2 && i == argc - 1)
				fatal("-l takes an argument");
			if (strlen(argv[i]) > 2)
				s = argv[i] + 2;
			else
				s = argv[++i];
			level = strtol(s, &endp, 10);
			if (*endp != '\0' || level <= 0)
				fatal("invalid level '%s'", s);
		} else if (strcmp(argv[i], "--gui") == 0) {
			if (i == argc - 1)
				fatal("--gui takes an argument");
			gui = argv[++i];
		} else if (strcmp(argv[i], "--size") == 0) {
			if (i == argc - 1)
				fatal("--size takes an argument");
			s = argv[++i];
			screensize = strtol(s, &endp, 10);
			if (*endp != '\0' || screensize <= 0)
				fatal("invalid screensize '%s'", s);
			if (screensize < SCREENSIZE)
				fatal("screensize must be larger than '%d'",
				       SCREENSIZE);
		} else if (strcmp(argv[1], "-v") == 0 ||
			   strcmp(argv[1], "--version") == 0)
		{
			printf ("XBill version 2.1\n\n");
			exit(0);
		} else if (strcmp(argv[1], "-h") == 0 ||
			   strcmp(argv[1], "--help") == 0)
		{
			printf("XBill version 2.1\n");
			printf("Options:\n");
			printf("-l <n>\tStart at level n.\n");
			printf("--gui <gui> \tUse a specific gui "
			       "(athena, motif, gtk)\n");
			printf("--size <size>\t\tUse a larger playing area.\n");
			printf("-v\t\tPrint version number and exit.\n");
			printf("-h\t\tPrint help and exit.\n");
			printf("All standard toolkit options are also ");
			printf("supported.\n\n");
			exit(0);
		}
	}
}

int
main(int argc, char **argv) {
	srand(time(NULL));
	parse_args(argc, argv);
	UI_initialize(gui, &argc, argv);
	UI_make_main_window(screensize);
	UI_graphics_init();
	UI_load_picture("logo", 0, &logo);
	UI_load_picture("icon", 0, &icon);
	UI_load_picture("about", 0, &about);
	draw_logo();
	UI_refresh();
	UI_make_dialogs(logo, icon, about);
	UI_set_icon(icon);

	Scorelist_read();
	Scorelist_update();

	UI_load_cursor("hand_up", CURSOR_SEP_MASK, &defaultcursor);
	UI_load_cursor("hand_down", CURSOR_SEP_MASK, &downcursor);
	UI_set_cursor(defaultcursor);

	Bill_load_pix();
	OS_load_pix();
	Computer_load_pix();
	Bucket_load_pix();
	Spark_load_pix();

	state = STATE_WAITING;
	if (level)
		Game_start(level);
	else
		UI_set_pausebutton(0);
	UI_main_loop();
	exit(0);
}
